### PR TITLE
feat: add resource/datasource `cloudavenue_edgegateway_nat_rule`

### PR DESCRIPTION
**Pull request by azrod**
### Description of your changes

If you submit change in the provider code, please make sure to:

- [x] Write or modify examples in `examples/` directory
- [x] Write or modify acceptance tests
- [x] Run `make generate` to ensure the doc was updated properly

### How has this code been tested

```TF_ACC=1 go test -v . -run TestAccNatRuleDataSource
=== RUN   TestAccNatRuleDataSource
--- PASS: TestAccNatRuleDataSource (162.31s)
PASS
ok      github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/tests/edgegw      162.677s
```

```source ~/.cloud_avenue.source && TF_ACC=1 go test -v . -run TestAccNATRuleResource -timeout 20m
=== RUN   TestAccNATRuleResource
--- PASS: TestAccNATRuleResource (730.44s)
PASS
ok      github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/tests/edgegw      730.746s
```

### CODE REVIEW

File: internal/provider/edgegw/nat_rule_schema.go [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+				Common: &schemaR.StringAttribute{
+					MarkdownDescription: "This represents the external port number or port range when doing DNAT port forwarding from external to internal. If not specify, all ports are translated",
+				},
+				Resource: &schemaR.StringAttribute{
+					Optional: true,
+					Validators: []validator.String{
+						fstringvalidator.RequireIfAttributeIsOneOf(path.MatchRoot("rule_type"), []attr.Value{types.StringValue("DNAT")}),
+						fstringvalidator.RequireIfAttributeIsOneOf(path.MatchRoot("rule_type"), []attr.Value{types.StringValue("NO_DNAT")}),
+					},
+				},
```
Comment (azrod) [inline internal/provider/edgegw/nat_rule_schema.go:197]:
```suggestion
					Validators: []validator.String{
						fstringvalidator.RequireIfAttributeIsOneOf(path.MatchRoot("rule_type"), []attr.Value{types.StringValue("DNAT"),types.StringValue("NO_DNAT")}),
					},
				},
```

File: internal/provider/edgegw/nat_rule_schema.go [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+			},
+			"snat_destination_address": superschema.SuperStringAttribute{
+				Common: &schemaR.StringAttribute{
+					MarkdownDescription: "The destination addresses to match in the SNAT Rule. This must be supplied as a single IP or Network CIDR. Providing no value for this field results in match with ANY destination network.",
+				},
+				Resource: &schemaR.StringAttribute{
+					Optional: true,
+					Validators: []validator.String{
+						fstringvalidator.RequireIfAttributeIsOneOf(path.MatchRoot("rule_type"), []attr.Value{types.StringValue("SNAT")}),
+						fstringvalidator.RequireIfAttributeIsOneOf(path.MatchRoot("rule_type"), []attr.Value{types.StringValue("NO_SNAT")}),
```
Comment (azrod) [inline internal/provider/edgegw/nat_rule_schema.go:210]:
```suggestion
						fstringvalidator.RequireIfAttributeIsOneOf(path.MatchRoot("rule_type"), []attr.Value{types.StringValue("SNAT"),types.StringValue("NO_SNAT")}),
```

File: internal/provider/edgegw/nat_rule_schema.go [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+			// 		MarkdownDescription: "Enable to have the address translation performed by this rule logged ",
+			// 	},
+			// 	Resource: &schemaR.BoolAttribute{
+			// 		Optional: true,
+			// 		Default:  booldefault.StaticBool(true),
+			// 	},
+			// 	DataSource: &schemaD.BoolAttribute{
+			// 		Computed: true,
+			// 	},
+			// },
```
Comment (azrod) [inline internal/provider/edgegw/nat_rule_schema.go:229]:
```suggestion
```

File: internal/provider/edgegw/nat_rule_schema.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+				Resource: &schemaR.StringAttribute{
+					Optional: true,
+					Validators: []validator.String{
+						fstringvalidator.OneOfWithDescription(
+							fstringvalidator.OneOfWithDescriptionValues{
+								Value:       "MATCH_INTERNAL_ADDRESS",
+								Description: "Applies firewall rules to the internal address of a NAT rule.",
+							},
+							fstringvalidator.OneOfWithDescriptionValues{
+								Value:       "MATCH_EXTERNAL_ADDRESST",
```
Comment (azrod) [inline internal/provider/edgegw/nat_rule_schema.go:253]:
```suggestion
								Value:       "MATCH_EXTERNAL_ADDRESS",
```

File: internal/provider/edgegw/nat_rule_types.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	DnatExternalPort   supertypes.StringValue `tfsdk:"dnat_external_port"`
+	EdgeGatewayID      supertypes.StringValue `tfsdk:"edge_gateway_id"`
+	EdgeGatewayName    supertypes.StringValue `tfsdk:"edge_gateway_name"`
+	Enabled            supertypes.BoolValue   `tfsdk:"enabled"`
+	ExternalAddress    supertypes.StringValue `tfsdk:"external_address"`
+	FirewallMatch      supertypes.StringValue `tfsdk:"firewall_match"`
+	ID                 supertypes.StringValue `tfsdk:"id"`
+	InternalAddress    supertypes.StringValue `tfsdk:"internal_address"`
+	// Option not available in CloudAvenue
+	// Logging                supertypes.BoolValue   `tfsdk:"logging"`
```
Comment (azrod) [inline internal/provider/edgegw/nat_rule_types.go:29]:
```suggestion
```

File: internal/provider/edgegw/nat_rule_types.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+		ExternalAddresses:        rm.ExternalAddress.Get(),
+		InternalAddresses:        rm.InternalAddress.Get(),
+		SnatDestinationAddresses: rm.SnatDestinationAddress.Get(),
+		DnatExternalPort:         rm.DnatExternalPort.Get(),
+		Type:                     rm.RuleType.Get(),
+		FirewallMatch:            rm.FirewallMatch.Get(),
+		Priority:                 utils.TakeIntPointer(int(rm.Priority.Get())),
+	}
+	// Get AppPortProfile
+	if rm.AppPortProfileID.IsKnown() {
```
Comment (azrod) [inline internal/provider/edgegw/nat_rule_types.go:120]: And `AppPortProfileName` ??

File: internal/tests/edgegw/nat_rule_resource_test.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+	external_address         = data.cloudavenue_publicips.example.public_ips[2].public_ip
+	internal_address         = "192.168.0.1"
+}
+`
+
+const testAccNATRuleResourceConfigDataSource = `
+data "cloudavenue_edgegateways" "example" {}
+
+data "cloudavenue_publicips" "example" {}
+`
```
Comment (azrod) [inline internal/tests/edgegw/nat_rule_resource_test.go:65]: Use EdgeGateway and PublicIP Unit Test for creating new resource

File: internal/tests/edgegw/nat_rule_resource_test.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+  
+	# Using primary_ip from edge gateway
+	external_address         = data.cloudavenue_publicips.example.public_ips[2].public_ip
+	internal_address         = "11.11.11.0/24"
+	snat_destination_address = "9.9.9.9"
+	
+	priority = 0
+}
+`
+
```
Comment (azrod) [inline internal/tests/edgegw/nat_rule_resource_test.go:83]: Add tests attributeNotSet

File: internal/tests/edgegw/nat_rule_resource_test.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+				Check:  natRuleReflexiveTestCheck(resourceName),
+			},
+			// Import State testing
+			{
+				// Import test
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+				ImportStateIdFunc: testAccNATRuleResourceImportStateIDFunc(resourceName),
+			},
```
Comment (azrod) [inline internal/tests/edgegw/nat_rule_resource_test.go:180]: Add one test for testing import edgeGatewayID and NatRuleID

File: templates/data-sources/edgegateway_nat_rule.tmpl
Comment (azrod) [inline templates/data-sources/edgegateway_nat_rule.tmpl:1]: The file name is not valid `templates/data-sources/edgegateway_nat_rule.tmpl` -> `templates/data-sources/edgegateway_nat_rule.md.tmpl`

File: docs/data-sources/edgegateway_nat_rule.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+<!-- schema generated by tfplugindocs -->
+## Schema
+
+### Required
+
+- `name` (String) The Name of the Nat Rule.
+
+### Optional
+
+- `description` (String) The Description of the Nat Rule.
```
Comment (azrod) [inline docs/data-sources/edgegateway_nat_rule.md:32]: ???

File: internal/provider/edgegw/nat_rule_schema.go (hunk 3 above)
Comment (dmicheneau) [reply internal/provider/edgegw/nat_rule_schema.go:229]: I let this comment in order to explain why this attribute is not implemented

File: internal/tests/edgegw/nat_rule_resource_test.go (hunk 9 above)
Comment (dmicheneau) [reply internal/tests/edgegw/nat_rule_resource_test.go:180]: Add test with invert attributes asked

File: docs/data-sources/edgegateway_nat_rule
[the remark this answers is not in the document]
Comment (dmicheneau) [reply docs/data-sources/edgegateway_nat_rule:1]: file deleted

File: .changelog/355.txt [hunk 11]
```diff
@@ -0,0 +1,6 @@
+```release-note:new-resource
+`resource/cloudavenue_edgegateway_nat_rules` - New resource to create a NAT Rule in edge gateway.
```
[the remark this answers is not in the document]
Comment (dmicheneau) [reply .changelog/355.txt]: Fix

File: docs/resources/edgegateway_nat_rule.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+# cloudavenue_edgegateway_nat_rule (Resource)
+
+The `cloudavenue_edgegateway_nat_rule` resource allows you to manage EdgeGateway NAT rule. To change the source IP address from a private to a public IP address, you create a source NAT (SNAT) rule. To change the destination IP address from a public to a private IP address, you create a destination NAT (DNAT) rule.
+
+## Example Usage
+
+```terraform
+resource "cloudavenue_edgegateway_nat_rule" "example" {
+  edge_gateway_name = "tn01e02ocb0006205spt103"
```
Comment (azrod) [inline docs/resources/edgegateway_nat_rule.md:16]:
```suggestion
  edge_gateway_name = "myEdgeName"
```

File: go.mod
Comment (azrod) [inline go.mod:1]: run `go mod tidy` for removing `github.com/k0kubun/pp` dep.

File: internal/provider/edgegw/nat_rule_schema.go [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+						),
+					},
+				},
+				DataSource: &schemaD.StringAttribute{
+					Computed: true,
+				},
+			},
+			"external_address": superschema.SuperStringAttribute{
+				Common: &schemaR.StringAttribute{
+					MarkdownDescription: "The external address for the NAT Rule. This must be supplied as a single IP or Network CIDR.For a DNAT rule, this is the external facing IP Address for incoming traffic. For an SNAT rule, this is the external facing IP Address for outgoing traffic. These IPs are typically allocated/suballocated IP Addresses on the Edge Gateway. For a REFLEXIVE rule, these are the external facing IPs.",
```
Comment (azrod) [inline internal/provider/edgegw/nat_rule_schema.go:148]:
```suggestion
					MarkdownDescription: "The external address for the NAT Rule. This must be supplied as a single IP or Network CIDR. For a DNAT rule, this is the external facing IP Address for incoming traffic. For an SNAT rule, this is the external facing IP Address for outgoing traffic. These IPs are typically allocated/suballocated IP Addresses on the Edge Gateway. For a REFLEXIVE rule, these are the external facing IPs.",
```

File: internal/provider/edgegw/nat_rule_schema.go [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+					Optional: true,
+					Validators: []validator.String{
+						fstringvalidator.RequireIfAttributeIsOneOf(path.MatchRoot("rule_type"), []attr.Value{types.StringValue("SNAT"), types.StringValue("NO_SNAT")}),
+						fstringvalidator.NullIfAttributeIsOneOf(path.MatchRoot("rule_type"), []attr.Value{types.StringValue("DNAT"), types.StringValue("NO_DNAT")}),
+					},
+				},
+				DataSource: &schemaD.StringAttribute{
+					Computed: true,
+				},
+			},
```
Comment (azrod) [inline internal/provider/edgegw/nat_rule_schema.go:211]: Same `dnat_external_port`

File: internal/provider/edgegw/nat_rule_types.go [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+		Name:                     rm.Name.Get(),
+		Description:              rm.Description.Get(),
+		Enabled:                  rm.Enabled.Get(),
+		ExternalAddresses:        rm.ExternalAddress.Get(),
+		InternalAddresses:        rm.InternalAddress.Get(),
+		SnatDestinationAddresses: rm.SnatDestinationAddress.Get(),
+		DnatExternalPort:         rm.DnatExternalPort.Get(),
+		Type:                     rm.RuleType.Get(),
+		FirewallMatch:            rm.FirewallMatch.Get(),
+		Priority:                 utils.TakeIntPointer(int(rm.Priority.Get())),
```
Comment (azrod) [inline internal/provider/edgegw/nat_rule_types.go:111]: In the attribute `Priority` the int pointer is used for differentiate `nil` and `0` value

Actually if Priority is not set `0` value is set and not `nil`

File: internal/tests/edgegw/nat_rule_resource_test.go
Comment (azrod) [inline internal/tests/edgegw/nat_rule_resource_test.go:1]: VDC Group is not tested

File: docs/resources/edgegateway_nat_rule.md (hunk 12 above)
Comment (dmicheneau) [reply docs/resources/edgegateway_nat_rule.md:16]: Fix

File: docs/data-sources/edgegateway_nat_rule.md [hunk 16]
```diff
@@ -0,0 +1,45 @@
+---
+page_title: "cloudavenue_edgegateway_nat_rule Data Source - cloudavenue"
+subcategory: "Edge Gateway (Tier-1)"
+description: |-
+  The cloudavenue_edgegateway_nat_rule data source allows you to retrieve information about an EdgeGateway NAT rule.
+---
+
+# cloudavenue_edgegateway_nat_rule (Data Source)
+
+The `cloudavenue_edgegateway_nat_rule` data source allows you to retrieve information about an EdgeGateway NAT rule.
```
[the remark this answers is not in the document]
Comment (dmicheneau) [reply docs/data-sources/edgegateway_nat_rule.md:10]: fix

File: go.mod
Comment (dmicheneau) [reply go.mod:1]: fix

File: internal/provider/edgegw/nat_rule_schema.go (hunk 13 above)
Comment (dmicheneau) [reply internal/provider/edgegw/nat_rule_schema.go:148]: Fix

File: internal/provider/edgegw/nat_rule_schema.go [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+					Optional: true,
+					Validators: []validator.String{
+						fstringvalidator.RequireIfAttributeIsOneOf(path.MatchRoot("rule_type"), []attr.Value{types.StringValue("DNAT"), types.StringValue("NO_DNAT")}),
+						fstringvalidator.NullIfAttributeIsOneOf(path.MatchRoot("rule_type"), []attr.Value{types.StringValue("SNAT"), types.StringValue("NO_SNAT")}),
+					},
+				},
+				DataSource: &schemaD.StringAttribute{
+					Computed: true,
+				},
+			},
```
[the remark this answers is not in the document]
Comment (dmicheneau) [reply internal/provider/edgegw/nat_rule_schema.go:196]: Fix: No need to require, need only to not set this attribute when SNAT like rule type is selected.

File: internal/provider/edgegw/nat_rule_schema.go (hunk 14 above)
Comment (dmicheneau) [reply internal/provider/edgegw/nat_rule_schema.go:211]: Fix, same also, i retired the require validator

File: internal/provider/edgegw/nat_rule_types.go (hunk 15 above)
Comment (dmicheneau) [reply internal/provider/edgegw/nat_rule_types.go:111]: This value is compute, and the default value return by api is value 0.

Comment (dmicheneau) [reply internal/provider/edgegw/nat_rule_types.go:111]: Now the value set in plan/state is unknow: Priority:               supertypes.NewInt64Unknown()

File: internal/tests/edgegw/nat_rule_resource_test.go
Comment (dmicheneau) [reply internal/tests/edgegw/nat_rule_resource_test.go:1]: test added

File: docs/resources/edgegateway_nat_rule.md [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+	name        = "example-reflexive"
+	rule_type   = "REFLEXIVE"
+	description = "description REFLEXIVE example"
+  
+	# Using primary_ip from edge gateway
+	external_address         = "89.32.25.10"
+	internal_address         = "192.168.0.1"
+
+  priority = 25
+}
```
Comment (azrod) [inline docs/resources/edgegateway_nat_rule.md:56]:
```suggestion
## Example NAT Rule SNAT

resource "cloudavenue_edgegateway_nat_rule" "example-snat" {
  edge_gateway_name = "myEdgeGateway"

  name            = "example-snat"
  rule_type     = "SNAT"
  description  = "description SNAT example"

  # Using primary_ip from edge gateway
  external_address                 = "89.32.25.10"
  internal_address                  = "11.11.11.0/24"
  snat_destination_address  = "8.8.8.8"

  priority = 10
}

## Example NAT Rule DNAT

resource "cloudavenue_edgegateway_nat_rule" "example-dnat" {
  edge_gateway_name = "myEdgeGateway"

  name           = "example-dnat"
  rule_type    = "DNAT"
  description = "description DNAT example"

 external_address         = "89.32.25.10"
 internal_address          = "4.11.11.11"
  
 dnat_external_port = "8080"
}

## Example NAT Rule reflexive

resource "cloudavenue_edgegateway_nat_rule" "example-reflexive" {
 edge_gateway_name = "myEdgeGateway"
  
 name            = "example-reflexive"
 rule_type     = "REFLEXIVE"
 description  = "description REFLEXIVE example"
  
 external_address         = "89.32.25.10"
 internal_address          = "192.168.0.1"

 priority = 25
}
```

**Pull request merged**